src/telemetry-tracing.ts

```diff
@@ -0,0 +1,492 @@
+ *
+ * Technically this is determined by the OpenTelemetry library, but
+ * in practice, it follows the W3C spec, so this should be the right
+ * one. The only thing we're using it for, anyway, is emptying user
+ * supplied attributes.
+ *
+ * @private
+ */
+export const modernAttributeName = 'googclient_traceparent';
+
+/**
+ * The old legacy attribute name.
+ *
+ * @private
+ */
+export const legacyAttributeName = 'googclient_OpenTelemetrySpanContext';
+
+export class SpanMaker {
+  static createPublisherSpan(message: PubsubMessage, topicName: string): Span {
+    const spanAttributes = {
+      // Add Opentelemetry semantic convention attributes to the span, based on:
+      // https://github.com/open-telemetry/opentelemetry-specification/blob/v1.1.0/specification/trace/semantic_conventions/messaging.md
+      [SemanticAttributes.MESSAGING_TEMP_DESTINATION]: false,
+      [SemanticAttributes.MESSAGING_SYSTEM]: 'pubsub',
+      [SemanticAttributes.MESSAGING_OPERATION]: 'send',
+      [SemanticAttributes.MESSAGING_DESTINATION]: topicName,
+      [SemanticAttributes.MESSAGING_DESTINATION_KIND]: 'topic',
+      [SemanticAttributes.MESSAGING_PROTOCOL]: 'pubsub',
+      [SemanticAttributes.MESSAGING_MESSAGE_PAYLOAD_SIZE_BYTES]:
+        message.data?.length,
+      'messaging.pubsub.ordering_key': message.orderingKey,
+    } as SpanAttributes;
+
+    const span: Span = getTracer().startSpan(`${topicName} send`, {
+      kind: SpanKind.PRODUCER,
+      attributes: spanAttributes,
+    });
+
+    return span;
+  }
+
+  static updatePublisherTopicName(span: Span, topicName: string) {
+    span.updateName(`${topicName} send`);
+    span.setAttribute(SemanticAttributes.MESSAGING_DESTINATION, topicName);
+  }
+
+  static createReceiveSpan(
+    message: MessageWithAttributes,
+    subName: string,
+    parent: Context | undefined
+  ): Span {
+    const name = `${subName} receive`;
+
+    // Mostly we want to keep the context IDs; the attributes and such
+    // are only something we do on the publish side.
+    if (context) {
+      return getTracer().startSpan(
+        name,
+        {
+          kind: SpanKind.CONSUMER,
+          attributes: {},
+        },
+        parent
+      );
+    } else {
+      return getTracer().startSpan(name, {
+        kind: SpanKind.CONSUMER,
+      });
+    }
+  }
+
+  static createChildSpan(
+    message: MessageWithAttributes,
+    name: string
+  ): Span | undefined {
+    const parent = message.telemetrySpan;
+    if (parent) {
+      return getTracer().startSpan(
+        name,
+        {
+          kind: SpanKind.INTERNAL,
+          attributes: {},
+        },
+        spanContextToContext(parent.spanContext())
+      );
+    } else {
+      return undefined;
+    }
+  }
+
+  static createPublishFlowSpan(message: PubsubMessage): Span | undefined {
+    return SpanMaker.createChildSpan(message, 'publisher flow control');
+  }
+
+  static createPublishBatchSpan(message: PubsubMessage): Span | undefined {
```

Comment:
we should also be recording how many messages are published in the same batch as an attribute